sphinx-docs/screenshot.py

```diff
@@ -0,0 +1,296 @@
+                         "start_url": "/securesync/login",
+                         "inputs": [{"#id_username":username},
+                                    {"#id_password":password},
+                                   ],
+                         "pages": [],
+                         "notes": "",
+                       }
+        if submit:
+            from_str_arg["inputs"].append({"<submit>":""})
+        from_str_arg["inputs"].append({"<slug>":self.filename})
+        if hasattr(self, "focus_selector"):
+            from_str_arg["focus"] = {}
+            from_str_arg["focus"]["selector"] = self.focus_selector
+            from_str_arg["focus"]["styles"] = FOCUS_CSS_STYLES
+            from_str_arg["notes"] = self.focus_annotation if hasattr(self, "focus_annotation") else ""
+        # Trying to import django.core.management.call_command gets you into some sort of import hell
+        # Apparently due to a circular import, according to Ben Bach.
+        self.env.screenshot_all_screenshots.append({
+            'docname':  self.env.docname,
+            'from_str_arg': from_str_arg,
+        })
+        self.arguments.append(os.path.join("/", "images", self.filename+".png"))
+        os.system("touch %s" % os.path.join("images", self.filename+".png"))
+        (image_node,) = Image.run(self)
+        return image_node
+
+    def _command_handler(self, commands):
+        if not self.url:
+            raise NotImplementedError("Please supply a url using the :url: option")
+        from_str_arg = { "users": [self.user_role],
+                         "slug": "",
+                         "start_url": self.url,
+                         "pages": [],
+                         "notes": "",
+                       }        
+        from_str_arg['inputs'] = reduce(lambda x,y: x+y, map(_cmd_to_inputs, commands), [])
+        from_str_arg["inputs"].append({"<slug>":self.filename})
+        if hasattr(self, "focus_selector"):
+            from_str_arg["focus"] = {}
+            from_str_arg["focus"]["selector"] = self.focus_selector
+            from_str_arg["focus"]["styles"] = FOCUS_CSS_STYLES
+            from_str_arg["notes"] = self.focus_annotation if hasattr(self, "focus_annotation") else ""
+        self.env.screenshot_all_screenshots.append({
+            'docname':  self.env.docname,
+            'from_str_arg': from_str_arg,
+        })
+        self.arguments.append(os.path.join("/", "images", self.filename+".png"))
+        os.system("touch %s" % os.path.join("images", self.filename+".png"))
+        (image_node,) = Image.run(self)
+        return image_node
+
+    # Add options to the image directive.
+    option_spec = Image.option_spec.copy()
+    option_spec['url'] = directives.unchanged
+    option_spec['user-role'] = _parse_user_role
+    option_spec['navigation-steps'] = _parse_nav_steps
+    option_spec['focus'] = _parse_focus
+
+    def run(self):
+        """ Returns list of nodes to appended as screenshot.
```

Comment:
Sure. Directives are required to implement this function and it's called at a certain point in the build process automatically. Briefly, sphinx reads the documents and builds an abstract document tree. Directives in general are parsed to return a list of nodes in the tree, which is what `run` is doing here. Then the nodes are actually built into a document depending on the builder chosen (i.e. html, pdf, latex)